Bounds are not always correctly set when setting twice a MediaStream to a HTMLMediaElement

youennf opened the pull request:
#### 38b13961316ce40f4b73db47c5c97fba86459c3b
<pre>
Bounds are not always correctly set when setting twice a MediaStream to a HTMLMediaElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=258723">https://bugs.webkit.org/show_bug.cgi?id=258723</a>
rdar://110753228

Reviewed by Jer Noble.

After double layer hosting, we were relying on MediaPlayerPrivateMediaStreamAVFObjC::setVideoInlineSizeFenced to be called
so that the AVSampleBufferDisplayLayer bound would be set.

This works fine initially but not when resetting srcObject to a new MediaStream.
In that case, we create a new SampleBufferDisplayLayer but the bounds may not be correctly initialized.
We are making sure to initialize both root and AVSampleBufferDisplayLayer bounds and positions.

They might still not be correctly set due to rotation.
This patch is thus refactoring the code so that LocalSampleBufferDisplayLayer will handle rotation itself.
This is feasible since it receives VideoFrames.
We remove sending affine transforms updates from WebProcess to GPUProcess.
Instead, whenever we receive a VideoFrame in LocalSampleBufferDisplayLayer, we compute the affine transform to use.
If there is a need for a new affine transform, we update the bounds of the AVSampleBufferDisplayLayer, which triggers
creating a new one currently.
This simplifies things as now, only the AVSampleBufferDisplayLayer has to get an affine transform and switch width and height depending on frame rotation.

To reduce flicker, we introduce m_isReconfiguring which will ensure we do not enqueue new video frames to an AVSampleBufferDisplayLayer
that will be removed soon.

We are also using inline video size instead of presentation size if possible to reduce reflowing.
There are still some undesired flicker/reflowing happening when rotation changes since both the media element bounds may change and the rotation.
And these steps are not done at once.

* LayoutTests/fast/mediastream/video-srcObject-set-twice-expected.html: Added.
* LayoutTests/fast/mediastream/video-srcObject-set-twice.html: Added.
* Source/WebCore/platform/graphics/avfoundation/SampleBufferDisplayLayer.h:
* Source/WebCore/platform/graphics/avfoundation/objc/LocalSampleBufferDisplayLayer.h:
* Source/WebCore/platform/graphics/avfoundation/objc/LocalSampleBufferDisplayLayer.mm:
(WebCore::LocalSampleBufferDisplayLayer::initialize):
(WebCore::LocalSampleBufferDisplayLayer::updateBoundsAndPosition):
(WebCore::LocalSampleBufferDisplayLayer::updateSampleLayerBoundsAndPosition):
(WebCore::videoTransformationMatrix):
(WebCore::LocalSampleBufferDisplayLayer::enqueueVideoFrame):
(WebCore::LocalSampleBufferDisplayLayer::enqueueBufferInternal):
(WebCore::LocalSampleBufferDisplayLayer::requestNotificationWhenReadyForVideoData):
(WebCore::LocalSampleBufferDisplayLayer::updateRootLayerAffineTransform): Deleted.
(WebCore::LocalSampleBufferDisplayLayer::updateAffineTransform): Deleted.
(WebCore::LocalSampleBufferDisplayLayer::setRootLayerBoundsAndPositions): Deleted.
(WebCore::LocalSampleBufferDisplayLayer::updateRootLayerBoundsAndPosition): Deleted.
(WebCore::LocalSampleBufferDisplayLayer::enqueueBuffer): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::enqueueVideoFrame):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::layersAreInitialized):
(WebCore::videoTransformationMatrix):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::rootLayerBoundsDidChange):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::setVideoInlineSizeFenced):
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.cpp:
(WebKit::RemoteSampleBufferDisplayLayer::updateBoundsAndPosition):
(WebKit::RemoteSampleBufferDisplayLayer::enqueueVideoFrame):
(WebKit::RemoteSampleBufferDisplayLayer::updateAffineTransform): Deleted.
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.h:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.messages.in:
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.cpp:
(WebKit::SampleBufferDisplayLayer::updateBoundsAndPosition):
(WebKit::SampleBufferDisplayLayer::updateAffineTransform): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.h:

Canonical link: <a href="https://commits.webkit.org/265708@main">https://commits.webkit.org/265708@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3bc71181e329120c66c21933b432e410514de687

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11695 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11895 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12262 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13339 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11140 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11714 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14283 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11875 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14006 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11859 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12701 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9922 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13760 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9993 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10618 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17751 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11070 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10772 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13945 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11172 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9219 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10349 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10497 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2806 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14629 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11029 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/graphics/avfoundation/objc/LocalSampleBufferDisplayLayer.mm:
```diff
@@ -182,16 +180,19 @@ static void runWithoutAnimations(const WTF::Function<void()>& function)
-    m_rootLayer.get().bounds = CGRectMake(0, 0, size.width(), size.height());
+    auto bounds = CGRectMake(0, 0, size.width(), size.height());
+    m_rootLayer.get().bounds = bounds;
+    m_rootLayer.get().position = { bounds.size.width / 2, bounds.size.height / 2 };
```
The defaults for `anchorPoint` will be different on different platforms, so it may be worthwhile to set these to `{ 0.5, 0.5 }` here.  Also keep in mind that `position` is relative to the layer's parent's bounds. Shouldn't be a problem here if the m_rootLayer's parent always has the same bounds as it does.

jernoble replied:
And same for m_sabpleBufferDisplayLayer, whose `anchorPoint` initialization was removed above.

youennf replied:
I will add back anchorPoint to m_sampleBufferDisplayLayer. Are you suggesting I do the same for the root layer even though we are not doing any transform to that layer?
Also which platform is not using `{ 0.5, 0.5 }`?

youennf committed:
Bounds are not always correctly set when setting twice a MediaStream to a HTMLMediaElement
https://bugs.webkit.org/show_bug.cgi?id=258723
rdar://110753228

Reviewed by Jer Noble.

After double layer hosting, we were relying on MediaPlayerPrivateMediaStreamAVFObjC::setVideoInlineSizeFenced to be called
so that the AVSampleBufferDisplayLayer bound would be set.

This works fine initially but not when resetting srcObject to a new MediaStream.
In that case, we create a new SampleBufferDisplayLayer but the bounds may not be correctly initialized.
We are making sure to initialize both root and AVSampleBufferDisplayLayer bounds and positions.

They might still not be correctly set due to rotation.
This patch is thus refactoring the code so that LocalSampleBufferDisplayLayer will handle rotation itself.
This is feasible since it receives VideoFrames.
We remove sending affine transforms updates from WebProcess to GPUProcess.
Instead, whenever we receive a VideoFrame in LocalSampleBufferDisplayLayer, we compute the affine transform to use.
If there is a need for a new affine transform, we update the bounds of the AVSampleBufferDisplayLayer, which triggers
creating a new one currently.
This simplifies things as now, only the AVSampleBufferDisplayLayer has to get an affine transform and switch width and height depending on frame rotation.

To reduce flicker, we introduce m_isReconfiguring which will ensure we do not enqueue new video frames to an AVSampleBufferDisplayLayer
that will be removed soon.

We are also using inline video size instead of presentation size if possible to reduce reflowing.
There are still some undesired flicker/reflowing happening when rotation changes since both the media element bounds may change and the rotation.
And these steps are not done at once.

* LayoutTests/fast/mediastream/video-srcObject-set-twice-expected.html: Added.
* LayoutTests/fast/mediastream/video-srcObject-set-twice.html: Added.
* Source/WebCore/platform/graphics/avfoundation/SampleBufferDisplayLayer.h:
* Source/WebCore/platform/graphics/avfoundation/objc/LocalSampleBufferDisplayLayer.h:
* Source/WebCore/platform/graphics/avfoundation/objc/LocalSampleBufferDisplayLayer.mm:
(WebCore::LocalSampleBufferDisplayLayer::initialize):
(WebCore::LocalSampleBufferDisplayLayer::updateBoundsAndPosition):
(WebCore::LocalSampleBufferDisplayLayer::updateSampleLayerBoundsAndPosition):
(WebCore::videoTransformationMatrix):
(WebCore::LocalSampleBufferDisplayLayer::enqueueVideoFrame):
(WebCore::LocalSampleBufferDisplayLayer::enqueueBufferInternal):
(WebCore::LocalSampleBufferDisplayLayer::requestNotificationWhenReadyForVideoData):
(WebCore::LocalSampleBufferDisplayLayer::updateRootLayerAffineTransform): Deleted.
(WebCore::LocalSampleBufferDisplayLayer::updateAffineTransform): Deleted.
(WebCore::LocalSampleBufferDisplayLayer::setRootLayerBoundsAndPositions): Deleted.
(WebCore::LocalSampleBufferDisplayLayer::updateRootLayerBoundsAndPosition): Deleted.
(WebCore::LocalSampleBufferDisplayLayer::enqueueBuffer): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::enqueueVideoFrame):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::layersAreInitialized):
(WebCore::videoTransformationMatrix):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::rootLayerBoundsDidChange):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::setVideoInlineSizeFenced):
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.cpp:
(WebKit::RemoteSampleBufferDisplayLayer::updateBoundsAndPosition):
(WebKit::RemoteSampleBufferDisplayLayer::enqueueVideoFrame):
(WebKit::RemoteSampleBufferDisplayLayer::updateAffineTransform): Deleted.
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.h:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.messages.in:
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.cpp:
(WebKit::SampleBufferDisplayLayer::updateBoundsAndPosition):
(WebKit::SampleBufferDisplayLayer::updateAffineTransform): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.h:

Canonical link: https://commits.webkit.org/265708@main
